source/game/kart/KartMove.cc

```diff
@@ -1593,6 +1624,150 @@ void KartMove::landTrick() {
+    forwardXZ.y = 0;
+    forwardXZ.normalise();
+    EGG::Vector3f local94 = m_cannonEntryOfs;
+    local94.y = 0;
+    local94.normalise();
+    m_speedRatioCapped = 1.0f;
+    EGG::Matrix34f cannonOrientation;
+    cannonOrientation.makeOrthonormalBasis(forward, EGG::Vector3f::ey);
+    EGG::Vector3f up = cannonOrientation.multVector33(EGG::Vector3f::ey);
+    m_smoothedUp = up;
+    m_up = up;
+
+    // exiting cannon
+    if ((forwardLength < 30.0f || local94.dot(forwardXZ) <= 0.0f) && state()->isInCannon()) {
+        state()->setInCannon(false);
+        state()->setSkipWheelCalc(false);
+        dynamics()->setIntVel(m_speed * m_cannonEntryOfs);
+        return;
+    }
+
+    cannonOrientation.mulRow33(0, EGG::Vector3f::ez);
+    cannonOrientation.mulRow33(1, EGG::Vector3f::ez);
+    m_speed = m_baseSpeed;
+    const auto *cannonPoint =
+            System::CourseMap::Instance()->getCannonPoint(state()->cannonPointId());
+    s16 cannonParameterIdx = cannonPoint->parameterIdx();
+    if (cannonParameterIdx < 1) {
+        cannonParameterIdx = 0;
+    } 
+    assert(static_cast<size_t>(cannonParameterIdx) < CANNON_PARAMETERS.size());
+    const auto cannonParams =  CANNON_PARAMETERS[cannonParameterIdx];
+    f32 newSpeed = cannonParams.speed;
+    if (forwardLength < cannonParams.decelFactor) {
+        f32 factor = std::max(0.0f, forwardLength / cannonParams.decelFactor);
+
+        newSpeed = cannonParams.endDecel;
+        if (newSpeed <= 0.0f) {
+            newSpeed = m_baseSpeed;
+        }
+
+        newSpeed += factor * (cannonParams.speed - newSpeed);
+        if ((cannonParams.endDecel > 0.0f) && (newSpeed < m_speed)) {
+            m_speed = newSpeed;
+        }
+    }
+
+    m_cannonProgress += m_cannonEntryOfs * newSpeed;
+
+    EGG::Vector3f newPos = EGG::Vector3f::zero;
+    if (cannonParams.height > 0.0f) {
+        const f32 fVar9 =
+                EGG::Mathf::SinFIdx((1.0f - (forwardLength / m_cannonEntryOfsLength)) * 180.0f * DEG2FIDX);
+        newPos = fVar9 * cannonParams.height * m_cannonOrthog;
+    }
+
+    dynamics()->setPos(m_cannonEntryPos + m_cannonProgress + newPos);
+    m_dir = m_cannonEntryOfs;
+    m_vel1Dir = m_cannonEntryOfs;
+
+    calcRotCannon(forward);
+
+    dynamics()->setExtVel(EGG::Vector3f::zero);
+}
+
+/// @addr{0x805855BC}
+void KartMove::calcRotCannon(const EGG::Vector3f &forward) {
+    EGG::Vector3f local48 = forward;
+    local48.normalise();
+    EGG::Vector3f local54 = bodyFront();
+    EGG::Vector3f local60 = local54 + ((local48 - local54) * 0.3f);
+    local54.normalise();
+    local60.normalise();
+    // also local70, localA8
+    EGG::Quatf local80;
+    local80.makeVectorRotation(local54, local60);
+    local80 *= dynamics()->fullRot();
+    local80.normalise();
+    EGG::Quatf localB8;
+    localB8.makeVectorRotation(local80.rotateVector(EGG::Vector3f::ey), smoothedUp());
+    const EGG::Quatf newRot = local80.slerpTo(localB8.multSwap(local80), 0.3f);
```

Comment:
Sorry for all of these, but vabold has established that `const`ing everything makes future expansion a little less clear, so we should avoid using these qualifiers for now.
```suggestion
    EGG::Quatf newRot = local80.slerpTo(localB8.multSwap(local80), 0.3f);
```